pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,193 @@
+		stateReplicationFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_state_replication_failed_total",
+			Help: "Number of times we have failed to replicate a state to other alertmanagers.",
+		}, []string{"key"}),
+	}
+
+	go s.handleBroadcasts()
+	return s
+}
+
+// AddState adds a new state that will be replicated using the ReplicationFunc. It returns a channel to which the client can broadcast messages of the state to be sent.
+func (s *state) AddState(key string, cs cluster.State, _ prometheus.Registerer) cluster.ClusterChannel {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.states[key] = cs
+
+	s.partialStateMergesTotal.WithLabelValues(key)
+	s.partialStateMergesFailed.WithLabelValues(key)
+	s.stateReplicationTotal.WithLabelValues(key)
+	s.stateReplicationFailed.WithLabelValues(key)
+
+	send := func(b []byte) {
+		s.msgc <- &clusterpb.Part{Key: key, Data: b}
+	}
+
+	return newStateChannel(key, send, s.logger)
+}
+
+// MergePartialState merges a received partial message with an internal state.
+func (s *state) MergePartialState(p *clusterpb.Part) error {
+	s.partialStateMergesTotal.WithLabelValues(p.Key).Inc()
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	st, ok := s.states[p.Key]
+	if !ok {
+		level.Error(s.logger).Log("msg", "state key not found while merging", "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return fmt.Errorf("unable to merge state, key '%s' not found while merging", p.Key)
+	}
+
+	if err := st.Merge(p.Data); err != nil {
+		level.Error(s.logger).Log("msg", "failed to merge state", "err", err, "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return errors.Wrap(err, "unable to merge state")
+	}
+
+	return nil
+}
+
+// The clusterWait function uses the position to determine how long should it wait before sending a notification.
+// In the Cortex case, we copy the function directly and use the manager to figure out which position we hold.
+func (s *state) Position() int { return 0 }
+
+// Settle waits until the alertmanagers are ready (and sets the appropriate internal state when it is).
+// The idea is that we don't want to start working" before we get a chance to know most of the notifications and/or silences.
+func (s *state) Settle(ctx context.Context, _ time.Duration) {
+	level.Info(s.logger).Log("msg", "Waiting for notification and silences to settle...")
+
+	// TODO: Make sure that the state is fully synchronised at this point.
+	// We can check other alertmanager(s) and explicitly ask them to propagate their state to us if available.
+	close(s.readyc)
+}
+
+// WaitReady is needed for the pipeline builder to know whenever we've settled and the state is up to date.
+func (s *state) WaitReady() {
+	//TODO: At the moment, we settle in a separate go-routine (see multitenant.go as we create the Peer) we should
+	// mimic that behaviour here once we have full state replication.
+	s.Settle(context.Background(), time.Second)
```

Comment:
`WaitReady` needs to be passed a context to solve this.

Reply:
Ditto from my comment below - Moreso, `WaitReady` as defined comes from upstream so we'd need to change that too.

Reply:
We don't need to copy `WaitReady` from upstream. We can define our own interface with what makes sense for us, and adapt upstream type to our interface via a wrapper.

Introducing context to upstream's `WaitReady` seems pretty straightforward too.

Reply:
DItto from my comment above, it is certainly possible but I would like to avoid it as part of this PR - The implementation of `WaitReady` and `Settle` for the new replication method are yet to be determined.